File: ex-11/ex-11_python/relogio.py
import time
from datetime import datetime


class Relogio:
    # Construtora com parâmetros
    def __init__(self, ano=None, mes=None, dia=None, hora=None, minutos=None, segundos=None, nanossegundos=None):
        if ano is not None:
            self.ano = ano
            self.mes = mes
            self.dia = dia
            self.hora = hora
            self.minutos = minutos
            self.segundos = segundos
            self.nanossegundos = nanossegundos
            self.tempoInicial = time.monotonic_ns()
        else:
            # Construtora com data/hora atuais
            # Obtém o tempo atual do sistema
            agora = datetime.now()

            self.ano = agora.year
            self.mes = agora.month
            self.dia = agora.day
            self.hora = agora.hour
            self.minutos = agora.minute
            self.segundos = agora.second
            # Obtém os milissegundos e converte para nanossegundos
            ms = agora.microsecond // 1000
            self.nanossegundos = ms * 1000000  # 1 ms = 10^6 ns

            self.tempoInicial = time.monotonic_ns()

    # Método getAgora() - retorna o tempo atual com precisão de nanossegundos
    def getAgora(self):
        # Calcula o tempo decorrido desde a criação do relógio
        tempoAtual = time.monotonic_ns()
        tempoDecorrido = tempoAtual - self.tempoInicial

        # Adiciona o tempo decorrido ao tempo original
        nsTotal = self.nanossegundos + tempoDecorrido
        segTotal = self.segundos
        minTotal = self.minutos
        horaTotal = self.hora
        diaTotal = self.dia

        # Ajusta nanossegundos (1 segundo = 10^9 nanossegundos)
        segTotal += nsTotal // 1000000000
        nsTotal = nsTotal % 1000000000

        # Ajusta segundos
        minTotal += segTotal // 60
        segTotal = segTotal % 60

        # Ajusta minutos
        horaTotal += minTotal // 60
        minTotal = minTotal % 60

        # Ajusta horas
        diaTotal += horaTotal // 24
        horaTotal = horaTotal % 24

        # Exibe o resultado
        print("{:04d}-{:02d}-{:02d} {:02d}:{:02d}:{:02d}.{:09d} ns".format(
            self.ano, self.mes, diaTotal, horaTotal, minTotal, segTotal, nsTotal))

    # Método para exibir o tempo inicial
    def exibirTempoInicial(self):
        print("{:04d}-{:02d}-{:02d} {:02d}:{:02d}:{:02d}.{:09d} ns".format(
            self.ano, self.mes, self.dia, self.hora, self.minutos, self.segundos, self.nanossegundos))

File: ex-11/ex-11_java/Relogio.java
import java.time.LocalDateTime;

public class Relogio {
    private int ano;
    private int mes;
    private int dia;
    private int hora;
    private int minutos;
    private int segundos;
    private long nanossegundos;

    // Armazena o tempo de criação do relógio
    private long tempoInicial;

    // Construtora com parâmetros
    public Relogio(int ano, int mes, int dia, int hora, int minutos, int segundos, long nanossegundos) {
        this.ano = ano;
        this.mes = mes;
        this.dia = dia;
        this.hora = hora;
        this.minutos = minutos;
        this.segundos = segundos;
        this.nanossegundos = nanossegundos;
        this.tempoInicial = System.nanoTime();
    }

    // Construtora com data/hora atuais
    public Relogio() {
        // Obtém o tempo atual do sistema
        LocalDateTime agora = LocalDateTime.now();

        this.ano = agora.getYear();
        this.mes = agora.getMonthValue();
        this.dia = agora.getDayOfMonth();
        this.hora = agora.getHour();
        this.minutos = agora.getMinute();
        this.segundos = agora.getSecond();
        // Obtém os milissegundos e converte para nanossegundos
        long ms = agora.getNano() / 1000000;
        this.nanossegundos = ms * 1000000; // 1 ms = 10^6 ns

        this.tempoInicial = System.nanoTime();
    }

    // Método getAgora() - retorna o tempo atual com precisão de nanossegundos
    public void getAgora() {
        // Calcula o tempo decorrido desde a criação do relógio
        long tempoAtual = System.nanoTime();
        long tempoDecorrido = tempoAtual - tempoInicial;

        // Adiciona o tempo decorrido ao tempo original
        long nsTotal = nanossegundos + tempoDecorrido;
        int segTotal = segundos;
        int minTotal = minutos;
        int horaTotal = hora;
        int diaTotal = dia;

        // Ajusta nanossegundos (1 segundo = 10^9 nanossegundos)
        segTotal += nsTotal / 1000000000;
        nsTotal = nsTotal % 1000000000;

        // Ajusta segundos
        minTotal += segTotal / 60;
        segTotal = segTotal % 60;

        // Ajusta minutos
        horaTotal += minTotal / 60;
        minTotal = minTotal % 60;

        // Ajusta horas
        diaTotal += horaTotal / 24;
        horaTotal = horaTotal % 24;

        // Exibe o resultado
        System.out.printf("%04d-%02d-%02d %02d:%02d:%02d.%09d ns%n",
                ano, mes, diaTotal, horaTotal, minTotal, segTotal, nsTotal);
    }

    // Método para exibir o tempo inicial
    public void exibirTempoInicial() {
        System.out.printf("%04d-%02d-%02d %02d:%02d:%02d.%09d ns%n",
                ano, mes, dia, hora, minutos, segundos, nanossegundos);
    }
}
